### PR TITLE
Reapply "Merge pull request #739 from petertseng/gzip-as-user" 

### PR DESCRIPTION
**Pull request by petertseng**
New commit means we are ready for this:

preparer: chmod TempDir on setup

As explained in code comments: If we extract as target user, we need to
allow them to see the tmpdir.

Original:

As we've seen time and time again, it's very difficult to get our custom
tar code right (see #116, #373, #374, #376, #623). We have decided
simply to call tar, dropping privileges to the RunAs user of the pod.

Using tar insulates us from any more edge cases we missed as long as
tar was the tool used to create the tarball, and dropping privileges
lets Unix permissions prevent any unintended system file modifications.

### CODE REVIEW

File: pkg/preparer/setup.go [hunk 1]
```diff
@@ -404,6 +404,14 @@ func New(preparerConfig *PreparerConfig, logger logging.Logger) (*Preparer, erro
 		return nil, util.Errorf("Could not create preparer pod directory: %s", err)
 	}
 
+	// Artifact files are downloaded to os.TempDir().
+	// Since we extract artifact files as target user, we must allow them to access the tmpdir.
+	// We expect that there is no sensitive information in TempDir, so 0755 is safe, though 0711 could be considered.
+	err = os.Chmod(os.TempDir(), 0755)
```
Comment (petertseng) [inline pkg/preparer/setup.go]: alternative: Perform this in the downloader.

Advantage: Action is closer to where it is necessary.
Disadvantage: Action performed multiple times.

Comment (petertseng) [inline pkg/preparer/setup.go]: Eh. I'm going to do it in the downloader. Interesting test interactions ("operation can't be nil"), if I have it here, I wonder what that means

Comment (petertseng) [inline pkg/preparer/setup.go]: OK, never mind, happens no matter where it goes.

File: pkg/preparer/setup.go [hunk 2]
```diff
@@ -404,6 +404,22 @@ func New(preparerConfig *PreparerConfig, logger logging.Logger) (*Preparer, erro
...
 
+	// Artifact files are downloaded to os.TempDir().
+	// Since we extract artifact files as target user, we must allow them to access the tmpdir.
+	// We expect that there is no sensitive information in TempDir, so 755 is safe, though 711 could be considered.
+	tmpDirStat, err := os.Stat(os.TempDir())
+	if err != nil {
+		return nil, util.Errorf("Could not stat tmpdir: %s", err)
+	}
+	mode := tmpDirStat.Mode()
+	if mode&0755 != 0755 {
```
Comment (petertseng) [inline pkg/preparer/setup.go]: this is my best guess as to why operation would be nil? If the mode was already 0755? Let's see...

**Comment by petertseng** [comment]
I honestly think https://travis-ci.org/square/p2/builds/223785567 (conditional, but always do 0755, don't preserve high bits) only succeeded because /tmp is already x755, so the test is inconclusive. I'll see what unconditional does though.

**Comment by petertseng** [comment]
OK, unconditional but keep high bitsfails, so conditional is needed.

File: pkg/preparer/setup.go [hunk 3]
```diff
@@ -404,6 +404,22 @@ func New(preparerConfig *PreparerConfig, logger logging.Logger) (*Preparer, erro
...
+		// keep whatever upper bit is there.
+		err = os.Chmod(os.TempDir(), (mode&07000)|0755)
```
Comment (petertseng) [inline pkg/preparer/setup.go]: wondering if anyone thinks this (keeping the highest... triad) is necessary. This was added after the most recent approval, would like a quick look at it.

File: pkg/preparer/setup.go (hunk 2 above)
Comment (petertseng) [inline pkg/preparer/setup.go]: Experimentally, this is the case, but I cannot reproduce this on my machine?!

File: pkg/preparer/setup.go (hunk 3 above)
Comment (mpuncel) [inline pkg/preparer/setup.go]: i doubt it will matter either way. can `(mode&07000)|0755` not just be written as `mode | 0755`?

Comment (mpuncel) [inline pkg/preparer/setup.go]: oh that would keep it executable

File: pkg/preparer/setup.go (hunk 2 above)
Comment (petertseng) [inline pkg/preparer/setup.go]: I'm going to leave a comment saying I can't reproduce it on any machine I have but Travis.

Perhaps Travis has special characteristics on its tmpdir.

**Comment by petertseng** [comment]
I've just realized what happened. "operation not permitted to be nil" is actually "expected **operation not permitted** to be nil". On Travis, we simply can't chmod the tmpdir.

**Pull request merged**